File: source/go/nested_cleanup.go
package main

import "fmt"

func inner() {
	defer fmt.Println("inner cleanup")
	fmt.Println("inner running")
}

func middle() {
	defer fmt.Println("middle cleanup")
	fmt.Println("middle running")
	inner()
}

func outer() {
	defer fmt.Println("outer cleanup")
	fmt.Println("outer running")
	middle()
}

func main() {
	outer()
}

File: source/javascript/nested_cleanup.js
// Nested cleanup (defer-like pattern via try/finally)
function inner() {
    try {
        console.log("inner running");
    } finally {
        console.log("inner cleanup");
    }
}

function middle() {
    try {
        console.log("middle running");
        inner();
    } finally {
        console.log("middle cleanup");
    }
}

function outer() {
    try {
        console.log("outer running");
        middle();
    } finally {
        console.log("outer cleanup");
    }
}

outer();
